src/shared/pkcs11-util.c

```diff
@@ -522,6 +522,296 @@ int pkcs11_token_find_x509_certificate(
+        case CKK_RSA: {
+                CK_ATTRIBUTE rsa_attributes[] = {
+                        { CKA_MODULUS,         NULL_PTR, 0 },
+                        { CKA_PUBLIC_EXPONENT, NULL_PTR, 0 }
+                };
+                _cleanup_free_ void *modulus = NULL;
+                _cleanup_free_ void *public_exponent = NULL;
+
+                rv = m->C_GetAttributeValue(session, object, rsa_attributes, ELEMENTSOF(rsa_attributes));
+                if (rv != CKR_OK)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                "Failed to get size of attributes of an RSA public key: %s", sym_p11_kit_strerror(rv));
+
+                if (rsa_attributes[0].ulValueLen != 0) {
+                        modulus = malloc(attributes[0].ulValueLen);
+                        if (!modulus)
+                                return log_oom();
+
+                        attributes[0].pValue = modulus;
+                } else
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "An RSA public key has empty CKA_MODULUS.");
+
+                if (rsa_attributes[1].ulValueLen != 0) {
+                        public_exponent = malloc(attributes[1].ulValueLen);
+                        if (!public_exponent)
+                                return log_oom();
+
+                        attributes[1].pValue = public_exponent;
+                } else
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "An RSA public key has empty CKA_PUBLIC_EXPONENT.");
+
+                rv = m->C_GetAttributeValue(session, object, rsa_attributes, ELEMENTSOF(rsa_attributes));
+                if (rv != CKR_OK)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                "Failed to get attributes of an RSA public key: %s", sym_p11_kit_strerror(rv));
+
+#if OPENSSL_VERSION_MAJOR >= 3
+                _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
+                if (!ctx)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to create an EVP_PKEY_CTX for RSA.");
+
+                if (EVP_PKEY_fromdata_init(ctx) != 1)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to init an EVP_PKEY_CTX for RSA.");
+
+                OSSL_PARAM rsa_params[] = {
+                        OSSL_PARAM_BN(OSSL_PKEY_PARAM_RSA_N, attributes[0].pValue, attributes[0].ulValueLen),
+                        OSSL_PARAM_BN(OSSL_PKEY_PARAM_RSA_E, attributes[1].pValue, attributes[1].ulValueLen),
+                        OSSL_PARAM_END
+                };
+
+                if (EVP_PKEY_fromdata(ctx, &pkey, EVP_PKEY_PUBLIC_KEY, rsa_params) != 1)
+                        return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "Failed to parse attributes of an RSA public key.");
+#else
+                BIGNUM *rsa_n = BN_bin2bn(attributes[0].pValue, attributes[0].ulValueLen, NULL);
```

Comment:
please use `_cleanup_(BN_freep)` on the BIGNUM object so that it is automatically freed when it goes out of scope

Reply:
(here and below)

Reply:
I see that `openssl-util.h` already have `rsa_pkey_from_n_e()`, so I will just switch to it.